waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+                     [{:op :test :path "/metadata/annotations/waiter~1app-status" :value "live"}
+                      {:op :test :path "/spec/replicas" :value replicas}
+                      {:op :replace :path "/spec/replicas" :value replicas'}]))
+
+(defn- get-replica-count
+  "Query the current replica count for the given Kubernetes object."
+  [{:keys [http-client] :as scheduler} replicaset-url]
+  (-> (api-request http-client replicaset-url)
+      (get-in [:spec :replicas])))
+
+(defmacro k8s-patch-with-retries
+  "Query the current replica count for the given Kubernetes object,
+   retrying a limited number of times in the event of an HTTP 409 conflict error."
+  [patch-cmd retry-condition retry-cmd]
+  `(let [patch-result# (ss/try+
+                         ~patch-cmd
+                         (catch [:status 409] _#
+                           ::conflict))]
+     (if (not= ::conflict patch-result#)
+       patch-result#
+       (if ~retry-condition
+         ~retry-cmd
+         (ss/throw+ {:status 409})))))
+
+(defn- scale-service-up-to
+  "Scale the number of instances for a given service to a specific number.
+   Only used for upward scaling. No-op if it would result in downward scaling."
+  [{:keys [api-server-url http-client max-conflict-retries] :as scheduler} service instances']
+  (let [replicaset-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (:namespace service)
+                         "/replicasets/"
+                         (:k8s-name service))]
+    (loop [attempt 1
+           instances (:instances service)]
+      (if (<= instances' instances)
+        (log/info "Skipping non-upward scale-up request on" (:id service)
+                  "from" instances "to" instances')
+        (k8s-patch-with-retries
+          (patch-object-replicas http-client replicaset-url instances instances')
+          (<= attempt max-conflict-retries)
+          (recur (inc attempt) (get-replica-count scheduler replicaset-url)))))))
+
+(defn- scale-service-by-delta
+  "Scale the number of instances for a given service by a given delta.
+   Can scale either upward (positive delta) or downward (negative delta)."
+  [{:keys [api-server-url http-client max-conflict-retries] :as scheduler} service instances-delta]
+  (let [replicaset-url (str api-server-url
+                            "/apis/extensions/v1beta1/namespaces/"
+                            (:namespace service)
+                            "/replicasets/"
+                            (:k8s-name service))]
+    (loop [attempt 1
+           instances (:instances service)]
+      (let [instances' (+ instances instances-delta)]
+        (k8s-patch-with-retries
+          (patch-object-replicas http-client replicaset-url instances instances')
+          (<= attempt max-conflict-retries)
+          (recur (inc attempt) (get-replica-count scheduler replicaset-url)))))))
+
+(defn- kill-service-instance
```

Comment:
What happens if any of the steps in the delete fail? E.g. deleting the instance succeeds but subsequent calls fail. The autoscaler will think the instance was not killed successfully and a subsequent request maybe assigned this killed instance causing the request to fail.

Reply:
You were right that this needed to be revisited. I added a patch to fix this. Now the operation returns success as long as the first `:delete` operation happens. I also updated the `pod->ServiceInstance` caller logic so that deleted pods (i.e., pods with the `deletionTimestamp` present) are no longer visible to Waiter.